crates/ra_ide_api/src/expand_macro.rs

```diff
@@ -0,0 +1,113 @@
+//! FIXME: write short doc here
+
+use crate::{db::RootDatabase, FilePosition};
+use hir::db::AstDatabase;
+use ra_db::SourceDatabase;
+use rustc_hash::FxHashMap;
+
+use ra_syntax::{
+    algo::{find_node_at_offset, replace_descendants},
+    ast::{self},
+    AstNode, NodeOrToken, SyntaxKind, SyntaxNode, WalkEvent,
+};
+
+fn insert_whitespaces(syn: SyntaxNode) -> String {
+    let mut res = String::new();
+
+    let mut token_iter = syn
+        .preorder_with_tokens()
+        .filter_map(|event| {
+            if let WalkEvent::Enter(NodeOrToken::Token(token)) = event {
+                Some(token)
+            } else {
+                None
+            }
+        })
+        .peekable();
+
+    while let Some(token) = token_iter.next() {
+        res += &token.text().to_string();
+        if token.kind().is_keyword()
+            || token.kind().is_literal()
+            || token.kind() == SyntaxKind::IDENT
+        {
+            if !token_iter.peek().map(|it| it.kind().is_punct()).unwrap_or(false) {
+                res += " ";
+            }
+        }
+    }
+
+    res
+}
+
+fn expand_macro_recur(
+    db: &RootDatabase,
+    source: hir::Source<&SyntaxNode>,
+    macro_call: &ast::MacroCall,
+) -> Option<SyntaxNode> {
+    let analyzer = hir::SourceAnalyzer::new(db, source, None);
+    let expansion = analyzer.expand(db, &macro_call)?;
+    let new_source = expansion.source(db);
+    let expanded: SyntaxNode = db.parse_or_expand(new_source.file_id)?;
+
+    let children = expanded.descendants().filter_map(ast::MacroCall::cast);
+    let mut replaces = FxHashMap::default();
+
+    for child in children.into_iter() {
+        let source = new_source.with_ast(source.ast);
+        let new_node = expand_macro_recur(db, source, &child)?;
+
+        replaces.insert(child.syntax().clone().into(), new_node.into());
+    }
+
+    Some(replace_descendants(&expanded, &replaces))
+}
+
+pub(crate) fn expand_macro(db: &RootDatabase, position: FilePosition) -> Option<(String, String)> {
```

Comment:
Let's move this function to the top of the file: it helps readability if public "entry-point" functions are near the begining.

Reply:
Let's also change the return type here, (and, more importantly, in `Analyzer::expand_macro`) to a struct instead of the tuple.

This is an *interface* function (that is, it is used far from definition), so it's important that, at the call site, the meaning of the return type is obvious. It's well worth to spend some time to define a dedicated struct.  

I suggest something like this:

```rust
pub struct ExpandedMacro {
    pub name: String,
    pub expansion: String,
}
```